hnn_core/parallel_backends.py

```diff
@@ -69,6 +81,220 @@ def _read_all_bytes(fd, chunk_size=4096):
+            proc_data_bytes = spliced_str_data.encode()
+
+            split_string = signal.split(':')
+            if len(split_string) > 1 and len(split_string[1]) > 0:
+                data_len = int(split_string[1])
+            else:
+                raise ValueError("Completion signal from child MPI process"
+                                 " did not contain data length.")
+
+            return proc_data_bytes, data_len
+
+    return data
+
+
+def run_subprocess(command, pickled_obj, timeout, *args, **kwargs):
+    """Run process asynchronously and communicate with it.
+
+    Parameters
+    ----------
+    command : list of str | str
+        Command to run as subprocess (see subprocess.Popen documentation).
+    pickled_obj : str
+        The pickled object to write to stdin after starting child process
+        with MPI command.
+    timeout : float
+        The number of seconds to wait after process ends.
+    *args, **kwargs : arguments
+        Additional arguments to pass to subprocess.Popen.
+
+    Returns
+    -------
+    proc : instance of Popen
+        The process instance.
+    proc_data_bytes : str
+        The processed data bytes returned by stdout.
+    data_len : int
+        The length of data.
+    """
+    proc_data_bytes = b''
+
+    # set up pairs of pipes to communicate with subprocess
+    (pipe_stdin_r, pipe_stdin_w) = os.pipe()
+    (pipe_stdout_r, pipe_stdout_w) = os.pipe()
+    (pipe_stderr_r, pipe_stderr_w) = os.pipe()
+
+    proc = Popen(command, stdin=pipe_stdin_r, stdout=pipe_stdout_w,
+                 stderr=pipe_stderr_w, *args, **kwargs)
+
+    # process will read stdin on startup
+    os.write(pipe_stdin_w, pickled_obj)
+
+    # signal that we are done writing pickled object
+    os.close(pipe_stdin_w)
+    os.close(pipe_stdin_r)
+
+    # create the selector instance and register all input events
+    # with read_stdout which will only echo to stdout
+    sel = selectors.DefaultSelector()
+    sel.register(pipe_stdout_r, selectors.EVENT_READ, _read_stdout)
+    sel.register(pipe_stderr_r, selectors.EVENT_READ, _read_stdout)
+
+    completed = False
+    total_time = 0
+    # wait timeout seconds after process stops to exit loop
+    while True:
+        if proc.poll() is not None:
+            if completed is True:
+                break
+            elif total_time > timeout:
+                # This is indicative of a failure. For debugging purposes.
+                warn(f"Timed out ({timeout}s) waiting for end of data after"
+                     " child process stopped")
+                break
+            else:
+                total_time += 1
+                sleep(1)
+
+        # wait for an event on the selector, timeout after 1s
+        events = sel.select(timeout=1)
+        for key, mask in events:
+            callback = key.data
+            callback_result = callback(key.fileobj, mask)
+            if callback.__name__ == '_read_stdout':
```

Comment:
explicitly using the name of the functions to make the code a bit more transparent